### PR TITLE
Modernize Objective-C layer and begin implementation of SDK layer

### PR DESCRIPTION
**Pull request by msmollin**
This PR brings in several commits from the `obj-c` that modernize, expand, and in general make it possible to package tangram into a framework that's a simple drag-and-drop into most systems. Additional streamlining is coming for build process / cmake integration, as well as expanded SDK functions, however we should integrate what's here for now into master since there are refactors that have occurred and new files introduced.

### CODE REVIEW

**Comment by msmollin** [comment]
Reviewers: When merging this, please leave the obj-c branch in tact. I have additional PRs open for this branch.

**Comment by blair1618** [comment]
I'm on this now - looking into massaging the build process to be a little simpler. 

I'm thinking we can use CMake to build the core tangram functionality into a fat, static library  using the iOS SDK and then commit an Xcode project that compiles the ViewController stuff and assembles a framework or demo application. Does that make sense?

**Comment by msmollin** [comment]
Hmmm that sounds like an ideal solution.

**Comment by msmollin** [comment]
What do you need to change here?

**Comment by blair1618** [comment]
I'm still messing around to figure that out haha

So far I'm changing `toolchains/iOS.cmake` to build `core` and `ios_platform.*` into a static library instead of an application bundle, which radically simplifies the CMake configuration. Then I guess I'll configure an Xcode project to locate that library and link to it. The part I haven't figured out yet is how to get the static library to re-build when a `core` file changes (or whether that's possible at all).

**Comment by blair1618** [comment]
Do you know whether an Xcode project can be configured to build another Xcode project as a "dependency"?

**Comment by msmollin** [comment]
Indeed it can be if you use a workspace. It's how cocoapods works its magic.

**Comment by blair1618** [comment]
Aha, workspaces! Ok this is what I'm thinking now, tell me if this makes sense:

We'll commit an Xcode workspace that has a project(s?) to build a framework and a demo app and also declares that those projects depend on another Xcode project, the one that builds the tangram static library. That static library Xcode project will be generated by CMake (and shouldn't do anything complicated). In this scenario, a user should only need to run a command to generate the static library project, then open the workspace and go wild from there.

**Comment by msmollin** [comment]
Hmmm that's crazy enough it could work, especially given it's very similar to how cocoapods generates their xcode project automatically.

**Comment by blair1618** [comment]
Well I haven't got the linking working yet, but this me hacking towards the structure I was talking about: https://github.com/tangrams/tangram-es/tree/matt/obj-c

**Comment by tallytalwar** [comment]
Hey @msmollin 

Can you update the Readme with this PR, to have the latest building process for ios and ios-simulator?

Thanks

**Comment by msmollin** [comment]
@tallytalwar done. Also rebased the branch to suck in master updates.

File: ios/src/TGMapViewController.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+struct TileID;
+@interface TGMapViewController : GLKViewController <UIGestureRecognizerDelegate>
+
+@property (assign, nonatomic) Tangram::Map* map;
+@property (assign, nonatomic) BOOL continuous;
+@property (weak, nonatomic) id<TangramGestureRecognizerDelegate> gestureDelegate;
+
+- (void)renderOnce;
+
+@end
```
Comment (blair1618) [inline ios/src/TGMapViewController.h]: Add trailing newline plz

File: ios/src/TGMapViewController.mm [hunk 2]
```diff
@@ -1,101 +1,97 @@
...
 
 
-@implementation ViewController
+@implementation TGMapViewController
 
 - (void)viewDidLoad
 {
     [super viewDidLoad];
-
+
```
Comment (blair1618) [inline ios/src/TGMapViewController.mm]: Trailing spaces should be trimmed, even on empty lines. I think you can configure Xcode to do this for you.

**Comment by blair1618** [comment]
Aside from my format nitpicking this looks good to me.

**Comment by msmollin** [comment]
@blair1618 newline added to the .h and I went a trimmed all the whitespace from the .mm file. Also updates the project file to be up2date with what's gone on in master.

**Comment by blair1618** [comment]
Thanks! LGTM.

**Comment by Songcn** [comment]
Hi, this branch is doing an excellent job to iOS developers. For a long time, tangram was difficult to be introduced into other iOS projects, developers had to copy a lot of resource files form tangram to their projects, not that easy as Android tangram. With this framework comes out, it will be more easy and friendly to iOS developers. Hope this pull request can be merged soon, and the framework can support x86 architecture soon. Thank you!

**Pull request merged**